File: algorithms/095.Unique_Binary_Search_Trees_II/Unique_Binary_Search_Trees_II.py
<<<<<<< HEAD
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
    
    def __repr__(self):
        if self:
            serial = []
            queue = [self]

            while queue:
                cur = queue[0]
                
                if cur:
                    serial.append(cur.val)
                    queue.append(cur.left)
                    queue.append(cur.right)
                else:
                    serial.append("#")
                
                queue = queue[1:]
                
            while serial[-1] == "#":
                serial.pop()
                
            return repr(serial)
                
        else:
            return None

class Solution:
    # @return a list of tree node
    def generateTrees(self, n):
        return self.generateTreesRecu(1, n)
    
    def generateTreesRecu(self, low, high):
        result = []
        if low > high:
            result.append(None)
        for i in xrange(low, high + 1):
            left = self.generateTreesRecu(low, i - 1)
            right = self.generateTreesRecu(i + 1, high)
            for j in left:
                for k in right:
                    cur = TreeNode(i)
                    cur.left = j
                    cur.right = k
                    result.append(cur)
        return result

if __name__ == "__main__":
    print Solution().generateTrees(3)
=======
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
    
    def __repr__(self):
        if self:
            serial = []
            queue = [self]

            while queue:
                cur = queue[0]
                
                if cur:
                    serial.append(cur.val)
                    queue.append(cur.left)
                    queue.append(cur.right)
                else:
                    serial.append("#")
                
                queue = queue[1:]
                
            while serial[-1] == "#":
                serial.pop()
                
            return repr(serial)
                
        else:
            return None

class Solution:
    # @return a list of tree node
    def generateTrees(self, n):
        return self.generateTreesRecu(1, n)
    
    def generateTreesRecu(self, low, high):
        result = []
        if low > high:
            result.append(None)
        for i in xrange(low, high + 1):
            left = self.generateTreesRecu(low, i - 1)
            right = self.generateTreesRecu(i + 1, high)
            for j in left:
                for k in right:
                    cur = TreeNode(i)
                    cur.left = j
                    cur.right = k
                    result.append(cur)
        return result

if __name__ == "__main__":
    print Solution().generateTrees(3)
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74

File: algorithms/095.Unique_Binary_Search_Trees_II/Unique_Binary_Search_Trees_II.cpp
#include <bits/stdc++.h>

using namespace std;


// Definition for binary tree
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    vector<TreeNode *> generate(int op, int ed) {
        vector<TreeNode *> ret;
        if (op > ed) {
            ret.push_back(NULL);
            return ret;
        }

        vector<TreeNode *> left, right;
        for (int k = op; k <= ed; ++k) {
            left = generate(op, k - 1);
            right = generate(k + 1, ed);

            for (auto &i : left)
                for (auto &j : right) {
                    TreeNode *root = new TreeNode(k);
                    root->left = i;
                    root->right = j;
                    ret.push_back(root);
                }
        }

        return ret;
    }
public:
    vector<TreeNode *> generateTrees(int n) {
        return generate(1, n);
    }
};
class Solution2 {
public:
    vector<TreeNode *> generateTrees(int n) {
       return createTree(1,n);
    }
    
    vector<TreeNode *> createTree(int start, int end)
    {
        vector<TreeNode *> results;
        if(start>end)
        {
            results.push_back(NULL);
            return results;
        }
        
        for(int k=start;k<=end;k++)
        {
            vector<TreeNode *> left = createTree(start,k-1);
            vector<TreeNode *> right = createTree(k+1,end);
            for(int i=0;i<left.size();i++)
            {
                for(int j=0;j<right.size();j++)
                {
                    TreeNode * root = new TreeNode(k);
                    root->left = left[i];
                    root->right = right[j];
                    results.push_back(root);
                }
            }
        }
        return results;
    }
};

int main() {

    return 0;
}

